### ISSUE TITLE
False positive on S1168 - returning null in LINQ lambda

### ISSUE THREAD

**Issue by Thieum** [SonarSource/sonar-dotnet#2461]
### Description

Regression of #761

Currently the rule [S1168](https://rules.sonarsource.com/csharp/RSPEC-1168) raises an issue when returning null from within a lambda declared inside of a method returning a collection.

### Repro steps

this code will demonstrate the false positive:

```
using System;
using System.Collections.Generic;
using System.Linq;

namespace ConsoleApp14
{
    static class Program
    {
        class ABC
        {
        }

        static void Main()
        {
            var test = Test();
            Console.WriteLine(test);
        }

        static IEnumerable<ABC> Test()
        {
            var list = new List<ABC>();

            return list.Select(o =>
            {
                if (o != null)
                {
                    return o;
                }
                return null;
            });
        }
    }
}
```

### Expected behavior

S1168 shouldn't be raised inside the scope of the lambda, as the return of the lambda is different than the one of the encapsulating method.

There could be legitimate case where the return of a lambda is a collection, where this rule could apply.

Both positive and negative results in lambda should be tested.

### Actual behavior

S1168 is raised inside the lambda considering the result of the encapsulating method instead of the result of the lambda.

### Known workarounds

ignore S1168 around the lambda expression.

### Related information

* SonarLint for Visual Studio 4.10.0.9867 / Sonar C# 7.14
* Visual Studio Enterprise 2019 16.1.3

**Comment by antoinebj** [comment]
Also encountered this

**Comment by costin-zaharia-sonarsource** [comment]
Yes, we also managed to reproduce the issue. We plan to make a hardening
sprint in the near future and will see how many of these we could fix.

Costin Zaharia | SonarSource

C# Developer https://sonarsource.com <http://sonarsource.com/>

Are you using SonarLint <http://www.sonarlint.org/> in your IDE?
